File: clients/ios/StellarChat/StellarChat/Models/AudioCache.swift
import Foundation

/// In-memory cache for decrypted voice message audio data.
/// Voice messages are small (< 10 MB) so disk caching is unnecessary.
final class AudioCache {
    static let shared = AudioCache()

    private let cache = NSCache<NSString, NSData>()

    private init() {
        cache.countLimit = 30
        cache.totalCostLimit = 30 * 1024 * 1024 // 30 MB
    }

    func data(for blobHash: String) -> Data? {
        cache.object(forKey: blobHash as NSString) as Data?
    }

    func store(_ data: Data, for blobHash: String) {
        cache.setObject(data as NSData, forKey: blobHash as NSString, cost: data.count)
    }
}

File: clients/android/StellarChat/app/src/main/java/chat/onym/android/blossom/AudioCache.kt
package chat.onym.android.blossom

/**
 * In-memory LRU cache for decrypted voice message audio data.
 * Voice messages are small (< 10 MB) so disk caching is unnecessary.
 */
object AudioCache {
    private val cache = object : LinkedHashMap<String, ByteArray>(16, 0.75f, true) {
        override fun removeEldestEntry(eldest: Map.Entry<String, ByteArray>): Boolean {
            return size > 30
        }
    }

    @Synchronized
    fun get(blobHash: String): ByteArray? = cache[blobHash]

    @Synchronized
    fun put(blobHash: String, data: ByteArray) {
        cache[blobHash] = data
    }
}
